index.rst

```diff
@@ -1,61 +1,181 @@
+data formats they would be expected to handle. In addition, it may be a useful
+tool for testing changes to the configuration of Rubin's alert brokers in the
+future.
 
-  Subsections
-  -----------
 
-  and
+High-level goals
+================
 
-  Subsubsections
-  ^^^^^^^^^^^^^^
+A simulator will help groups who are developing community brokers which
+republish Rubin's alert stream. It will provide a realistic benchmark for the
+load they should expect to receive, and will act as a harness for them to run
+tests.
 
-  To add images, add the image file (png, svg or jpeg preferred) to the
-  _static/ directory. The reST syntax for adding the image is
+The simulator should:
 
-  .. figure:: /_static/filename.ext
-     :name: fig-label
+ - Provide a realistic stream of alerts, both in terms of volume and in content.
+ - Be simple to install on realistic broker hardware
+ - Be well-documented
+ - Provide tools to make it clear whether a consumer is keeping up with the
+   stream volume
 
-     Caption text.
+Implementation
+==============
 
-   Run: ``make html`` and ``open _build/html/index.html`` to preview your work.
-   See the README at https://github.com/lsst-sqre/lsst-technote-bootstrap or
-   this repo's README for more info.
+The simulator will be provided as a set of Docker containers and a
+``docker-compose`` file for coordinating those containers. To run the simulator,
+the user spins up the docker containers, and then runs a command which starts
+the flow of data. The data comes from a directory with serialized alert data on
+disk, and is paced into the containerized Kafka broker to match the expected
+rate for Rubin. The data will be repeated indefinitely, with timestamps
+adjusted, so that long-term behavior can be observed, even if only a few minutes
+of source data is available.
 
-   Feel free to delete this instructional comment.
+Alert data will be serialized using the latest Avro alert event schema. The Avro
+schema specification used will be provided along with the containers.
+
+Docker
+------
 
-:tocdepth: 1
+These containers will be provided:
 
-.. Please do not modify tocdepth; will be fixed when a new Sphinx theme is shipped.
+ - A Kafka container, which runs the broker
+ - A Zookeeper container, which the broker connects to for backend state management
+ - A monitoring container, which provides a web-based UI verify that the stream
+   is being fully handled
+ - An injector container, which hosts a command that sends data to the broker
 
-.. sectnum::
+The Kafka broker and monitoring containers will be addressable from the host
+running the containers so that the user can consume the stream and check metrics
+easily.
+
+Injector and Data Rates
+-----------------------
 
-.. TODO: Delete the note below before merging new content to the master branch.
+The injector sends data to the broker. It should consume a directory on disk,
+and will repeat the data it finds there indefinitely.
```

Comment:
I  think this is fine, but this is the area about which I'm most fuzzy in my thinking.

Specifically, I wonder about things like:

- Repeating indefinitely means that at some point things like `alertID`s and `diaObjectId`s will start repeating. Is that a problem?
- Do we need to think about the format in which things are serialized? Is the data on disk in Avro format already, or is there some more efficient dump format that we'll be distributing? (It's probably premature to worry about this.)
- Will it be possible for broker authors to download meaningful subsets of the data? The full 10e6 alert dump in MongoDB form is a few hundred GB as I recall; folks might want to just get 10 minutes of data, or subset by source class, or something.

Reply:
I suspect @ebellm is really the person to weigh in here — he could help clarify requirements. But off the top of my head:

- Consider making a cmd line option to switch between “loop forever” mode, which will reuse `alertId`s etc, or “single pass” mode, which will just send all the data once and then stop.
- Eric can comment on whether it's actually useful to provide all of these simulated alerts to consumers — I believe his current thinking is that it might be more effective just to provide a relatively small number of precursor events instead. However, if it *is* useful, my first suggestion would be to ask in #dm-infrastructure for hosting ideas.

Reply:
- I agree that a "loop forever by duplicating data" and a "single pass" switch would be helpful.  While in principle we could edit all the relevant ids on the fly to reuse data it seems like a huge headache--let's wait and see if there's demand for that from the broker teams.
- I think we should plan to provide a large bulk sample of alerts to consumers if (and when) we have it--I suspect it will be instructive for all concerned, especially as we start planning for the Data Previews during commissioning.

Reply:
I really like the command line switch mechanism too, let's do that.

I'm going to avoid going into detail in this particular doc on how we provide the big dump of data, but I'll ask in #dm-infrastructure.

Maybe we can include a small snapshot of data directly in the repository, perhaps one wave of ~10,000 alerts. That should be less than a gigabyte, probably a lot less if we compress it.

Reply:
@ctslater and @ktlim suggested serving the data blobs directly off of `$HOME/public_html` on a `lsst-dev` server. We can put hundreds of gigabytes up there and there are no concerns about hogging bandwidth on the outbound flows.